Emby.Server.Implementations/SyncPlay/SyncPlayManager.cs

```diff
@@ -41,338 +44,415 @@ public class SyncPlayManager : ISyncPlayManager, IDisposable
-                    return;
-                }
+                return new List<GroupInfoDto>();
+            }
 
-                group.SessionLeave(session, cancellationToken);
+            var user = _userManager.GetUserById(session.UserId);
+            List<GroupInfoDto> list = new List<GroupInfoDto>();
 
-                if (group.IsGroupEmpty())
+            // Locking required to access list of groups.
+            lock (_groupsLock)
+            {
+                foreach (var group in _groups.Values)
                 {
-                    _logger.LogInformation("LeaveGroup: removing empty group {0}.", group.GetGroupId());
-                    _groups.Remove(group.GetGroupId(), out _);
+                    // Locking required as group is not thread-safe.
+                    lock (group)
+                    {
+                        if (group.HasAccessToPlayQueue(user))
+                        {
+                            list.Add(group.GetInfo());
+                        }
+                    }
                 }
             }
+
+            return list;
         }
 
         /// <inheritdoc />
-        public List<GroupInfoView> ListGroups(SessionInfo session, Guid filterItemId)
+        public void HandleRequest(SessionInfo session, IGroupPlaybackRequest request, CancellationToken cancellationToken)
         {
-            var user = _userManager.GetUserById(session.UserId);
+            if (!IsRequestValid(session, request))
+            {
+                return;
+            }
 
-            if (user.SyncPlayAccess == SyncPlayAccess.None)
+            IGroupController group;
+            lock (_mapsLock)
             {
-                return new List<GroupInfoView>();
+                group = FindJoinedGroup(session);
             }
 
-            // Filter by item if requested
-            if (!filterItemId.Equals(Guid.Empty))
+            if (group == null)
             {
-                return _groups.Values.Where(
-                    group => group.GetPlayingItemId().Equals(filterItemId) && HasAccessToItem(user, group.GetPlayingItemId())).Select(
-                    group => group.GetInfo()).ToList();
+                _logger.LogWarning("Session {SessionId} does not belong to any group.", session.Id);
+
+                var error = new GroupUpdate<string>(Guid.Empty, GroupUpdateType.NotInGroup, string.Empty);
+                _sessionManager.SendSyncPlayGroupUpdate(session, error, CancellationToken.None);
+                return;
             }
-            else
+
+            // Group lock required as GroupController is not thread-safe.
+            lock (group)
             {
-                // Otherwise show all available groups
-                return _groups.Values.Where(
-                    group => HasAccessToItem(user, group.GetPlayingItemId())).Select(
-                    group => group.GetInfo()).ToList();
+                group.HandleRequest(session, request, cancellationToken);
```

Comment:
This one's also about the **Visitor pattern**. Instead of having multiple if-else/switch statements to dispatch the request to the right state, overloading is used.

Actually, maybe this one's more about `SyncPlayManager` not having to care about the underneath type of the request.

But I hope you get the idea. Should I write a comment to make this easier to understand?